File: src/main/java/aurocosh/divinefavor/common/util/UtilCoordinates.java
package aurocosh.divinefavor.common.util;

import net.minecraft.block.state.IBlockState;
import net.minecraft.init.Blocks;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import java.util.*;
import java.util.function.Function;
import java.util.function.Predicate;

public class UtilCoordinates {
    public static List<BlockPos> getRandomNeighbours(BlockPos start, int neighbourCount, final int minNextNode, final int maxNextNode, int cycleLimit, Predicate<BlockPos> predicate) {
        List<BlockPos> selectedNodes = new ArrayList<>();
        Queue<BlockPos> nodesToVisit = new ArrayDeque<>();
        Set<BlockPos> plannedNodes = new HashSet<>();
        Set<BlockPos> visitedNodes = new HashSet<>();
        nodesToVisit.add(start);

        while (!nodesToVisit.isEmpty() && cycleLimit-- > 0 && neighbourCount-- > 0) {
            BlockPos nextNode = nodesToVisit.remove();
            plannedNodes.remove(nextNode);
            visitedNodes.add(nextNode);
            selectedNodes.add(nextNode);

            if (!predicate.test(nextNode))
                continue;

            int neighboursToAdd = UtilRandom.nextInt(minNextNode, maxNextNode);
            List<BlockPos> neighbours = UtilBlockPos.getNeighbours(nextNode);
            List<BlockPos> newNodes = new ArrayList<>();
            for (BlockPos neighbour : neighbours)
                if (!visitedNodes.contains(neighbour) && !plannedNodes.contains(neighbour))
                    newNodes.add(neighbour);
            newNodes = UtilRandom.selectRandom(newNodes, neighboursToAdd);
            nodesToVisit.addAll(newNodes);
            plannedNodes.addAll(newNodes);
        }
        return selectedNodes;
    }

    public static List<BlockPos> getBlocksInSphere(final BlockPos center, final int radius) {
        final List<BlockPos> cubeCoordinates = getBlocksInRadius(center, radius, radius, radius);
        final int radiusSq = radius * radius;
        return UtilList.select(cubeCoordinates, pos -> center.distanceSq(pos) <= radiusSq);
    }

    public static List<BlockPos> getBlocksInRadius(BlockPos center, int radiusX, int radiusY, int radiusZ) {
        List<BlockPos> coordinates = new ArrayList<>();
        for (int x = -radiusX; x <= radiusX; x++)
            for (int y = -radiusY; y <= radiusY; y++)
                for (int z = -radiusZ; z <= radiusZ; z++)
                    coordinates.add(center.add(x, y, z));
        return coordinates;
    }

    public static BlockPos getRandomNeighbourSafe(BlockPos center, int xRadius, int yRadius, int zRadius) {
        BlockPos destination = UtilCoordinates.getRandomNeighbour(center, xRadius, yRadius, zRadius);
        return new BlockPos(destination.getX(), Math.max(5, destination.getY()), destination.getZ());
    }

    public static BlockPos getRandomNeighbour(BlockPos center, int xRadius, int yRadius, int zRadius) {
        int xShift = UtilRandom.nextInt(-xRadius, xRadius);
        int yShift = UtilRandom.nextInt(-yRadius, yRadius);
        int zShift = UtilRandom.nextInt(-zRadius, zRadius);
        return center.add(xShift, yShift, zShift);
    }

    public static BlockPos getRandomBlockInRange(BlockPos center, int radius, int limit, Predicate<BlockPos> predicate) {
        BlockPos blockPos = getRandomNeighbour(center, radius, radius, radius);
        if (predicate.test(blockPos))
            return blockPos;

        blockPos = findPosition(center, limit, predicate, BlockPos::down);
        return blockPos != null ? blockPos : findPosition(center, limit, predicate, BlockPos::up);
    }

    public static BlockPos findPlaceToStand(BlockPos start, World world, int limit) {
        BlockPos pos = findPlaceToStandBelow(start.up(), world, limit, true);
        if (pos != null)
            return pos;
        return findPlaceToTeleportAbove(start.down(), world, limit);
    }

    public static BlockPos findPlaceToTeleport(BlockPos start, World world, EnumFacing facing, int limit, boolean needPlaceToStand) {
        BlockPos pos = start;
        while (limit-- > 0) {
            pos = pos.offset(facing);
            if (!world.isAirBlock(pos) || !world.isAirBlock(pos.up()))
                continue;
            if (!needPlaceToStand)
                return pos;
            IBlockState state = world.getBlockState(pos.down());
            if (state.getBlock() == Blocks.BEDROCK)
                return null;
            if (state.isSideSolid(world, pos.down(), EnumFacing.UP))
                return pos;
        }
        return null;
    }

    public static BlockPos findPlaceToTeleportAbove(BlockPos start, World world, int limit) {
        BlockPos pos = start;
        boolean previousSecondIsAir = false;
        boolean previousOneIsAir = false;
        while (limit-- > 0) {
            pos = pos.up();
            if (previousOneIsAir && previousSecondIsAir)
                return pos.down();

            IBlockState state = world.getBlockState(pos);
            if (state.getBlock() == Blocks.BEDROCK)
                return null;

            previousSecondIsAir = previousOneIsAir;
            previousOneIsAir = world.isAirBlock(pos);
        }
        return null;
    }

    public static BlockPos findPlaceToStandBelow(BlockPos start, World world, int limit, boolean needPlaceToStand) {
        BlockPos pos = start;
        boolean previousSecondIsAir = false;
        boolean previousOneIsAir = false;
        while (limit-- > 0) {
            pos = pos.down();
            IBlockState state = world.getBlockState(pos);
            if (previousOneIsAir && previousSecondIsAir && (!needPlaceToStand || state.isSideSolid(world, pos, EnumFacing.UP)))
                return pos.up();
            if (state.getBlock() == Blocks.BEDROCK)
                return null;

            previousSecondIsAir = previousOneIsAir;
            previousOneIsAir = world.isAirBlock(pos);
        }
        return null;
    }

    public static BlockPos findPosition(final BlockPos start, int limit, Predicate<BlockPos> predicate, Function<BlockPos, BlockPos> nextPosFunction) {
        BlockPos pos = start;
        while (limit-- > 0) {
            if (predicate.test(pos))
                return pos;
            pos = nextPosFunction.apply(pos);
        }
        return null;
    }

    public static List<BlockPos> floodFill(List<BlockPos> start, List<BlockPos> expansionDirs, Predicate<BlockPos> predicate, int limit) {
        Queue<BlockPos> expansionFront = new ArrayDeque<>(start);
        Set<BlockPos> explored = new HashSet<>(start);
        List<BlockPos> result = new ArrayList<>();

        while (expansionFront.size() > 0 && result.size() < limit) {
            BlockPos nextPos = expansionFront.remove();
            if (!predicate.test(nextPos))
                continue;
            result.add(nextPos);
            for (BlockPos expansionDir : expansionDirs) {
                BlockPos neighbour = nextPos.add(expansionDir);
                if (!explored.contains(neighbour)) {
                    expansionFront.add(neighbour);
                    explored.add(neighbour);
                }
            }
        }
        return result;
    }
}

File: src/main/java/aurocosh/divinefavor/common/util/UtilCoordinates.kt
package aurocosh.divinefavor.common.util

import aurocosh.divinefavor.common.constants.BlockPosConstants
import aurocosh.divinefavor.common.lib.extensions.selectRandom
import net.minecraft.init.Blocks
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d
import net.minecraft.world.World
import java.util.*

object UtilCoordinates {
    fun getRandomNeighbours(start: BlockPos, neighbourCount: Int, minNextNode: Int, maxNextNode: Int, cycleLimit: Int, predicate: (BlockPos) -> Boolean): List<BlockPos> {
        var neighbourCounter = neighbourCount
        var cycleCounter = cycleLimit
        val selectedNodes = ArrayList<BlockPos>()
        val nodesToVisit = ArrayDeque<BlockPos>()
        val plannedNodes = HashSet<BlockPos>()
        val visitedNodes = HashSet<BlockPos>()
        nodesToVisit.add(start)

        while (!nodesToVisit.isEmpty() && cycleCounter-- > 0 && neighbourCounter-- > 0) {
            val nextNode = nodesToVisit.remove()
            plannedNodes.remove(nextNode)
            visitedNodes.add(nextNode)
            selectedNodes.add(nextNode)

            if (!predicate.invoke(nextNode))
                continue

            val neighboursToAdd = UtilRandom.nextInt(minNextNode, maxNextNode)
            val randomNewNodes = BlockPosConstants.DIRECT_NEIGHBOURS
                    .map(nextNode::add)
                    .filter { node -> !visitedNodes.contains(node) && !plannedNodes.contains(node) }
                    .selectRandom(neighboursToAdd)
            nodesToVisit.addAll(randomNewNodes)
            plannedNodes.addAll(randomNewNodes)
        }
        return selectedNodes
    }

    fun getSphereOutline(center: BlockPos, radiusInternal: Int, radiusExternal: Int): List<BlockPos> {
        val cubeCoordinates = getBlocksInSphere(center, radiusExternal)
        val radiusSq = radiusInternal * radiusInternal
        return UtilList.select(cubeCoordinates) { pos -> center.distanceSq(pos) >= radiusSq }
    }

    fun getBlocksInSphere(center: BlockPos, radius: Int): List<BlockPos> {
        val cubeCoordinates = getBlocksInRadius(center, radius, radius, radius)
        val radiusSq = radius * radius
        return UtilList.select(cubeCoordinates) { pos -> center.distanceSq(pos) <= radiusSq }
    }

    fun getBlocksInRadius(center: BlockPos, radiusX: Int, radiusY: Int, radiusZ: Int): List<BlockPos> {
        val coordinates = ArrayList<BlockPos>()
        for (x in -radiusX..radiusX)
            for (y in -radiusY..radiusY)
                for (z in -radiusZ..radiusZ)
                    coordinates.add(center.add(x, y, z))
        return coordinates
    }

    fun getRandomNeighbourSafe(center: BlockPos, xRadius: Int, yRadius: Int, zRadius: Int): BlockPos {
        val destination = getRandomNeighbour(center, xRadius, yRadius, zRadius)
        return BlockPos(destination.x, Math.max(5, destination.y), destination.z)
    }

    fun getRandomNeighbour(center: BlockPos, xRadius: Int, yRadius: Int, zRadius: Int): BlockPos {
        val xShift = UtilRandom.nextInt(-xRadius, xRadius)
        val yShift = UtilRandom.nextInt(-yRadius, yRadius)
        val zShift = UtilRandom.nextInt(-zRadius, zRadius)
        return center.add(xShift, yShift, zShift)
    }

    fun getRandomBlockInRange(center: BlockPos, radius: Int, limit: Int, predicate: (BlockPos) -> Boolean): BlockPos? {
        val blockPos: BlockPos = getRandomNeighbour(center, radius, radius, radius)
        if (predicate.invoke(blockPos))
            return blockPos

        return findPosition(center, limit, predicate, { it.down() }) ?: findPosition(center, limit, predicate, { it.up() })
    }

    fun findPlaceToStand(start: BlockPos, world: World, limit: Int): BlockPos? {
        val pos = findPlaceToStandBelow(start.up(), world, limit, true)
        return pos ?: findPlaceToTeleportAbove(start.down(), world, limit)
    }

    fun findPlaceToTeleport(start: BlockPos, world: World, facing: EnumFacing, limit: Int, needPlaceToStand: Boolean): BlockPos? {
        var limitCounter = limit
        var pos = start
        while (limitCounter-- > 0) {
            pos = pos.offset(facing)
            if (!world.isAirBlock(pos) || !world.isAirBlock(pos.up()))
                continue
            if (!needPlaceToStand)
                return pos
            val state = world.getBlockState(pos.down())
            if (state.block === Blocks.BEDROCK)
                return null
            if (state.isSideSolid(world, pos.down(), EnumFacing.UP))
                return pos
        }
        return null
    }

    fun findPlaceToTeleportAbove(start: BlockPos, world: World, limit: Int): BlockPos? {
        var limitCounter = limit
        var pos = start
        var previousSecondIsAir = false
        var previousOneIsAir = false
        while (limitCounter-- > 0) {
            pos = pos.up()
            if (previousOneIsAir && previousSecondIsAir)
                return pos.down()

            val state = world.getBlockState(pos)
            if (state.block === Blocks.BEDROCK)
                return null

            previousSecondIsAir = previousOneIsAir
            previousOneIsAir = world.isAirBlock(pos)
        }
        return null
    }

    fun findPlaceToStandBelow(start: BlockPos, world: World, limit: Int, needPlaceToStand: Boolean): BlockPos? {
        var limitCounter = limit
        var pos = start
        var previousSecondIsAir = false
        var previousOneIsAir = false
        while (limitCounter-- > 0) {
            pos = pos.down()
            val state = world.getBlockState(pos)
            if (previousOneIsAir && previousSecondIsAir && (!needPlaceToStand || state.isSideSolid(world, pos, EnumFacing.UP)))
                return pos.up()
            if (state.block === Blocks.BEDROCK)
                return null

            previousSecondIsAir = previousOneIsAir
            previousOneIsAir = world.isAirBlock(pos)
        }
        return null
    }

    fun findPosition(start: BlockPos, limit: Int, predicate: (BlockPos) -> Boolean, nextPosFunction: (BlockPos) -> BlockPos): BlockPos? {
        var limitCounter = limit
        var pos = start
        while (limitCounter-- > 0) {
            if (predicate.invoke(pos))
                return pos
            pos = nextPosFunction.invoke(pos)
        }
        return null
    }

    fun floodFill(start: List<BlockPos>, expansionDirs: List<BlockPos>, predicate: (BlockPos) -> Boolean, limit: Int): List<BlockPos> {
        val expansionFront = ArrayDeque(start)
        val explored = HashSet(start)
        val result = ArrayList<BlockPos>()

        while (expansionFront.size > 0 && result.size < limit) {
            val nextPos = expansionFront.remove()
            if (!predicate.invoke(nextPos))
                continue
            result.add(nextPos)
            for (expansionDir in expansionDirs) {
                val neighbour = nextPos.add(expansionDir)
                if (!explored.contains(neighbour)) {
                    expansionFront.add(neighbour)
                    explored.add(neighbour)
                }
            }
        }
        return result
    }

    fun getBoundingBox(origin: BlockPos, radius: Double): AxisAlignedBB {
        return AxisAlignedBB(origin.x - radius, origin.y - radius, origin.z - radius, origin.x + radius, origin.y + radius, origin.z + radius)
    }

    fun getBoundingBox(origin: Vec3d, radius: Double): AxisAlignedBB {
        return AxisAlignedBB(origin.x - radius, origin.y - radius, origin.z - radius, origin.x + radius, origin.y + radius, origin.z + radius)
    }
}
